ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -802,48 +802,67 @@ func (r *RayServiceReconciler) reconcileIngress(ctx context.Context, rayServiceI
 }
 
 func (r *RayServiceReconciler) reconcileServices(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService, rayClusterInstance *rayv1alpha1.RayCluster, serviceType common.ServiceType) error {
-	// Creat Service Struct.
-	var raySvc *corev1.Service
+	r.Log.Info(
+		"reconcileServices", "serviceType", serviceType,
+		"RayService name", rayServiceInstance.Name, "RayService namespace", rayServiceInstance.Namespace,
+	)
+
+	var newSvc *corev1.Service
 	var err error
-	if serviceType == common.HeadService {
-		raySvc, err = common.BuildHeadServiceForRayService(*rayServiceInstance, *rayClusterInstance)
-	} else if serviceType == common.ServingService {
-		raySvc, err = common.BuildServeServiceForRayService(*rayServiceInstance, *rayClusterInstance)
+
+	switch serviceType {
+	case common.HeadService:
+		newSvc, err = common.BuildHeadServiceForRayService(*rayServiceInstance, *rayClusterInstance)
+	case common.ServingService:
+		newSvc, err = common.BuildServeServiceForRayService(*rayServiceInstance, *rayClusterInstance)
+	default:
+		return fmt.Errorf("unknown service type %v", serviceType)
 	}
 
+	// TODO: This seems to be useless because BuildHeadServiceForRayService and BuildServeServiceForRayService will not return error.
 	if err != nil {
 		return err
 	}
-	raySvc.Name = utils.CheckName(raySvc.Name)
+	r.Log.Info("reconcileServices", "newSvc", newSvc)
 
-	// Get Service instance.
-	rayService := &corev1.Service{}
-	err = r.Get(ctx, client.ObjectKey{Name: raySvc.Name, Namespace: rayServiceInstance.Namespace}, rayService)
+	// Retrieve the Service from the Kubernetes cluster with the name and namespace.
+	oldSvc := &corev1.Service{}
+	err = r.Get(ctx, client.ObjectKey{Name: newSvc.Name, Namespace: rayServiceInstance.Namespace}, oldSvc)
 
 	if err == nil {
-		// Update Service
-		rayService.Spec = raySvc.Spec
-		r.Log.V(1).Info("reconcileServices update service")
-		if updateErr := r.Update(ctx, rayService); updateErr != nil {
-			r.Log.Error(updateErr, "raySvc Update error!", "raySvc.Error", updateErr)
+		// Only update the service if the RayCluster switches.
+		if newSvc.Spec.Selector[common.RayClusterLabelKey] == oldSvc.Spec.Selector[common.RayClusterLabelKey] {
+			r.Log.Info(fmt.Sprintf("RayCluster %v's %v has already exists, skip Update", newSvc.Spec.Selector[common.RayClusterLabelKey], serviceType))
+			return nil
+		}
+
+		// ClusterIP is immutable. Starting from Kubernetes v1.21.5, if the new service does not specify a ClusterIP,
+		// Kubernetes will assign the ClusterIP of the old service to the new one. However, to maintain compatibility
+		// with older versions of Kubernetes, we need to assign the ClusterIP here.
+		if newSvc.Spec.ClusterIP == "" {
+			newSvc.Spec.ClusterIP = oldSvc.Spec.ClusterIP
+		}
+		oldSvc.Spec = *newSvc.Spec.DeepCopy()
```

Comment:
We need to consider the updates of `ObjectMeta` in subsequent PRs. We currently only take `Spec` into consideration.

Reply:
Added https://github.com/ray-project/kuberay/pull/1110/commits/3119e679e2ff22459961482f2c79bf53efd2b41f